src/gz/bufread.rs

```diff
@@ -34,87 +34,118 @@ fn bad_header() -> io::Error {
+                }
+                r.part.state = GzHeaderParsingState::Filename;
+            }
+            GzHeaderParsingState::Filename => {
+                if r.part.flg & FNAME != 0 {
+                    if None == r.part.header.filename {
+                        let vec = Vec::new();
+                        r.part.header.filename = Some(vec);
+                    };
+                    if let Some(ref mut b) = r.part.header.filename {
+                        for byte in r.reader.bytes() {
+                            let byte = byte?;
+                            r.crc.update(&[byte]);
+                            if byte == 0 {
+                                break;
+                            }
+                            b.push(byte);
+                        }
+                    }
+                }
+                r.part.state = GzHeaderParsingState::Comment;
+            }
+            GzHeaderParsingState::Comment => {
+                if r.part.flg & FCOMMENT != 0 {
+                    if None == r.part.header.comment {
+                        let vec = Vec::new();
+                        r.part.header.comment = Some(vec);
+                    };
+                    if let Some(ref mut b) = r.part.header.comment {
+                        for byte in r.reader.bytes() {
+                            let byte = byte?;
+                            r.crc.update(&[byte]);
+                            if byte == 0 {
+                                break;
+                            }
+                            b.push(byte);
+                        }
+                    }
+                }
+                r.part.state = GzHeaderParsingState::Crc;
+            }
+            GzHeaderParsingState::Crc => {
+                if r.part.flg & FHCRC != 0 {
+                    let stored_crc = read_le_u16_once(r)?;
+                    let calced_crc = r.crc.sum() as u16;
+                    if stored_crc != calced_crc {
+                        return Err(corrupt());
+                    }
+                }
+                return Ok(());
             }
-            b.push(byte);
         }
-        Some(b)
-    } else {
-        None
+    }
+}
+
+pub(crate) fn read_gz_header<R: Read>(r: &mut R) -> io::Result<GzHeader> {
+    let mut part = GzHeaderPartial::new();
+
+    let result = {
+        let mut reader = Buffer::new(&mut part, r);
+        read_gz_header_part(&mut reader)
     };
 
-    if flg & FHCRC != 0 {
-        let calced_crc = crc_reader.crc().sum() as u16;
-        let stored_crc = read_le_u16(&mut crc_reader)?;
-        if calced_crc != stored_crc {
-            return Err(corrupt());
+    match result {
+        Ok(()) => {
+            return Ok(part.take_header());
         }
-    }
-
-    Ok(GzHeader {
-        extra: extra,
-        filename: filename,
-        comment: comment,
-        operating_system: os,
-        mtime: mtime,
-    })
+        Err(err) => {
+            return Err(err);
```

Comment:
Could this use `?` instead of explicit returns?

Reply:
indeed nicer